src/ott/neural/models/base_solver.py

```diff
@@ -0,0 +1,485 @@
+      eta_predictions = apply_fn_eta({"params": params_eta}, x, condition)
+      return (
+          optax.l2_loss(eta_predictions[:, 0], a).mean() +
+          optax.l2_loss(jnp.mean(eta_predictions) - expectation_reweighting),
+          eta_predictions,
+      )
+
+    def loss_b_fn(
+        params_xi: Optional[jnp.ndarray],
+        apply_fn_xi: Callable[[Dict[str, jnp.ndarray], jnp.ndarray],
+                              jnp.ndarray],
+        x: jnp.ndarray,
+        condition: Optional[jnp.ndarray],
+        b: jnp.ndarray,
+        expectation_reweighting: float,
+    ) -> Tuple[float, jnp.ndarray]:
+      xi_predictions = apply_fn_xi({"params": params_xi}, x, condition)
+      return (
+          optax.l2_loss(xi_predictions[:, 0], b).mean() +
+          optax.l2_loss(jnp.mean(xi_predictions) - expectation_reweighting),
+          xi_predictions,
+      )
+
+    @jax.jit
+    def step_fn(
+        source: jnp.ndarray,
+        target: jnp.ndarray,
+        condition: Optional[jnp.ndarray],
+        a: jnp.ndarray,
+        b: jnp.ndarray,
+        state_eta: Optional[train_state.TrainState] = None,
+        state_xi: Optional[train_state.TrainState] = None,
+        *,
+        is_training: bool = True,
+    ):
+      if state_eta is not None:
+        grad_a_fn = jax.value_and_grad(loss_a_fn, argnums=0, has_aux=True)
+        (loss_a, eta_predictions), grads_eta = grad_a_fn(
+            state_eta.params,
+            state_eta.apply_fn,
+            source,
+            condition,
+            a * len(a),
+            jnp.sum(b),
+        )
+        new_state_eta = state_eta.apply_gradients(
+            grads=grads_eta
+        ) if is_training else None
+
+      else:
+        new_state_eta = eta_predictions = loss_a = None
+      if state_xi is not None:
+        grad_b_fn = jax.value_and_grad(loss_b_fn, argnums=0, has_aux=True)
+        (loss_b, xi_predictions), grads_xi = grad_b_fn(
+            state_xi.params,
+            state_xi.apply_fn,
+            target,
+            condition,
+            b * len(b),
+            jnp.sum(a),
+        )
+        new_state_xi = state_xi.apply_gradients(
+            grads=grads_xi
+        ) if is_training else None
+      else:
+        new_state_xi = xi_predictions = loss_b = None
+
+      return (
+          new_state_eta, new_state_xi, eta_predictions, xi_predictions, loss_a,
+          loss_b
+      )
+
+    return step_fn
+
+  def evaluate_eta(
+      self,
+      source: jnp.ndarray,
+      condition: Optional[jnp.ndarray] = None
+  ) -> jnp.ndarray:
+    """Evaluate the left learnt rescaling factor.
+
+    Args:
+      source: Samples from the source distribution to evaluate rescaling
+        function on.
+      condition: Condition belonging to the samples in the source distribution.
+
+    Returns:
+      Learnt left rescaling factors.
+    """
+    if self.state_eta is None:
+      raise ValueError("The left rescaling factor was not parameterized.")
```

Comment:
Question: instead of throwing an error, wouldn't it make more sense to just return an array of 1s?